sherpa/astro/instrument.py

```diff
@@ -1400,48 +1413,242 @@ def create_non_delta_rmf(rmflo, rmfhi, fname, offset=1,
     # Set up the delta-function response.
     # TODO: should f_chan start at startchan?
     #
-    nchans = rmflo.size
+    # Is this a square matrix or not?
+    if e_min is None:
+        nchans = rmflo.size
+    else:
+        nchans = e_min.size
 
     n_grp, f_chan, n_chan, matrix = calc_grp_chan_matrix(fname)
+    return DataRMF(name, detchans=nchans, energ_lo=rmflo,
+                   energ_hi=rmfhi, n_grp=n_grp,
+                   n_chan=n_chan, f_chan=f_chan,
+                   matrix=matrix, offset=offset,
+                   e_min=e_min, e_max=e_max,
+                   ethresh=ethresh, header=header)
+
+
+def calc_grp_chan_matrix(fname: str) -> tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray, numpy.ndarray]:
+    """Read in an image and convert it to RMF components.
+
+    For an image containing a RMF - so X axis is for channels and Y
+    axis is for the energy axis - extract the needed data to create a
+    DataRMF object (modulo knowledge of the channel or energy grids).
+
+    Parameters
+    ----------
+    fname : str
+       The file name.
+
+    Returns
+    -------
+    n_grp, f_chan, n_chan, matrix : (ndarray, ndarray, ndarray, ndarray)
+       Needed to create a DataRMF to match fname.
+
+    """
+
+    if TYPE_CHECKING:
+        # Assume we have an I/O backend
+        assert io.backend is not None
+
+    # TODO: this could use the WCS info to create the channel and
+    # energy arrays, at least for files created by rmfimg.
+    #
+    data, _ = io.backend.get_image_data(fname)
+    return matrix_to_rmf(data["y"])
+
+
+def matrix_to_rmf(matrix: numpy.ndarray) -> tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray, numpy.ndarray]:
+    """Convert a matrix (2D image) to RMF components.
+
+    .. versionadded:: 4.16.0
+
+    Parameters
+    ----------
+    matrix : ndarray
+       A 2D matrix (X axis is channels, Y axis is energy).
+
+    Returns
+    -------
+    n_grp, f_chan, n_chan, matrix : (ndarray, ndarray, ndarray, ndarray)
+       Needed to create a DataRMF to match matrix.
+
+    """
+
+    if matrix.ndim != 2:
+        raise ValueError(f"matrix must be 2D, not {matrix.ndim}D")
+
+    n_grp1: list[int] = []
+    n_chan1: list[int] = []
+    f_chan1: list[int] = []
+    for row in matrix > 0:
+        flag = numpy.hstack([[0], row, [0]])
+        diffs = numpy.diff(flag, n=1)
+        starts, = numpy.where(diffs > 0)
+        ends, = numpy.where(diffs < 0)
+        n_chan1.extend(ends - starts)
+        f_chan1.extend(starts + 1)
```

Comment:
At this point we don't know what the desired offset is - we have no metadata sent in with the ndarray to indicate what the offset/TLMIN value should be - so I think it makes sense to pick one value and use that. 

So I think like 1490 says "we anoint offset=1" as the one true representation here. I guess it could be an optional argument to the routine, but I'd like to wait on adding optional arguments until we work out we need them. Which we may well do, I am probably too close to the API to be objective now!